File: youzi_v2/schemas/maritime_alerts.py
from pydantic import BaseModel, ConfigDict, Field


class MaritimeAlertCounts(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    arriving_soon: int = Field(0, validation_alias="arrivingSoon")
    departing_soon: int = Field(0, validation_alias="departingSoon")
    arrived: int = Field(0, validation_alias="arrived")
    in_transit: int = Field(0, validation_alias="inTransit")
    planned: int = Field(0, validation_alias="planned")
    unknown: int = Field(0, validation_alias="unknown")
    port_arriving_soon: int = Field(0, validation_alias="portArrivingSoon")
    port_departing_soon: int = Field(0, validation_alias="portDepartingSoon")


class MaritimeAlertShipment(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    shipment_no: str = Field(validation_alias="shipmentNo")
    customer: str | None = None
    vessel_voyage: str | None = Field(None, validation_alias="vesselVoyage")
    maritime_status: str = Field(validation_alias="maritimeStatus")
    maritime_status_label: str = Field(validation_alias="maritimeStatusLabel")
    eta: str | None = None
    etd: str | None = None
    destination_port_code: str | None = Field(None, validation_alias="destinationPortCode")


class MaritimeAlertPortCall(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    voyage_id: str = Field(validation_alias="voyageId")
    vessel_voyage: str = Field(validation_alias="vesselVoyage")
    port_name: str = Field(validation_alias="portName")
    sequence: int
    status: str
    status_label: str = Field(validation_alias="statusLabel")
    eta: str | None = None
    etd: str | None = None


class PortArrivalNotification(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    subscription_id: str = Field(validation_alias="subscriptionId")
    port_call_id: str = Field(validation_alias="portCallId")
    voyage_id: str = Field(validation_alias="voyageId")
    port_name: str = Field(validation_alias="portName")
    vessel_voyage: str = Field(validation_alias="vesselVoyage")
    ata: str
    created_at: str = Field(validation_alias="createdAt")
    read_at: str | None = Field(None, validation_alias="readAt")


class MaritimeAlertsOverview(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    generated_at: str = Field(validation_alias="generatedAt")
    counts: MaritimeAlertCounts
    urgent_shipments: list[MaritimeAlertShipment] = Field(
        default_factory=list, validation_alias="urgentShipments"
    )
    urgent_port_calls: list[MaritimeAlertPortCall] = Field(
        default_factory=list, validation_alias="urgentPortCalls"
    )
    eta_arriving_soon_port_calls: list[MaritimeAlertPortCall] = Field(
        default_factory=list, validation_alias="etaArrivingSoonPortCalls"
    )
    eta_arriving_soon_shipments: list[MaritimeAlertShipment] = Field(
        default_factory=list, validation_alias="etaArrivingSoonShipments"
    )
    voyages_with_alerts: list[dict] = Field(
        default_factory=list, validation_alias="voyagesWithAlerts"
    )
    unconfigured_vessel_voyages: list[dict] = Field(
        default_factory=list, validation_alias="unconfiguredVesselVoyages"
    )
    total_scanned: int = Field(0, validation_alias="totalScanned")
    port_arrival_notifications: list[PortArrivalNotification] = Field(
        default_factory=list, validation_alias="portArrivalNotifications"
    )

File: youzi_v2/frontend/src/api/maritimeAlerts.ts
import { api } from '@/api/client'
import type { MaritimeStatus } from '@/types/vesselSchedule'

export interface MaritimeAlertCounts {
  arrivingSoon: number
  departingSoon: number
  arrived: number
  inTransit: number
  planned: number
  unknown: number
  portArrivingSoon: number
  portDepartingSoon: number
}

export interface MaritimeAlertShipment {
  shipmentNo: string
  customer: string | null
  vesselVoyage: string | null
  maritimeStatus: MaritimeStatus
  maritimeStatusLabel: string
  eta: string | null
  etd: string | null
  destinationPortCode: string | null
}

export interface MaritimeAlertPortCall {
  voyageId: string
  vesselVoyage: string
  portName: string
  sequence: number
  status: MaritimeStatus
  statusLabel: string
  eta: string | null
  etd: string | null
}

export interface VoyageAlertBrief {
  voyageId: string
  vesselVoyage: string
  arrivingSoonPorts: number
  departingSoonPorts: number
}

export interface UnconfiguredVesselVoyage {
  vesselVoyage: string
  shipmentCount: number
}

export interface MaritimeAlertsOverview {
  generatedAt: string
  counts: MaritimeAlertCounts
  urgentShipments: MaritimeAlertShipment[]
  urgentPortCalls: MaritimeAlertPortCall[]
  /** 挂靠港 ETA 三天内到港（无需订阅） */
  etaArrivingSoonPortCalls: MaritimeAlertPortCall[]
  /** 运单 ETA 三天内到港 */
  etaArrivingSoonShipments: MaritimeAlertShipment[]
  voyagesWithAlerts: VoyageAlertBrief[]
  unconfiguredVesselVoyages: UnconfiguredVesselVoyage[]
  totalScanned: number
  portArrivalNotifications: PortArrivalNotification[]
  shipmentArrivalNotifications: ShipmentArrivalNotification[]
}

export interface PortArrivalNotification {
  id: string
  subscriptionId: string
  portCallId: string
  voyageId: string
  portName: string
  vesselVoyage: string
  ata: string
  createdAt: string
  readAt: string | null
}

export interface ShipmentArrivalNotification {
  id: string
  subscriptionId: string
  shipmentId: string
  shipmentNo: string
  customer: string | null
  vesselVoyage: string
  destinationPortCode: string
  trackingSource: 'internal' | 'carrier' | 'arrival' | string
  latestTime: string
  latestDesc: string
  createdAt: string
  readAt: string | null
}

export async function getMaritimeAlertsOverview(): Promise<MaritimeAlertsOverview> {
  return api<MaritimeAlertsOverview>('/api/v1/maritime-alerts/overview')
}

export async function markPortArrivalNotificationRead(notificationId: string): Promise<void> {
  await api(`/api/v1/maritime-alerts/port-arrivals/${encodeURIComponent(notificationId)}/read`, {
    method: 'POST',
  })
}

export async function markShipmentArrivalNotificationRead(notificationId: string): Promise<void> {
  await api(
    `/api/v1/maritime-alerts/shipment-arrivals/${encodeURIComponent(notificationId)}/read`,
    { method: 'POST' },
  )
}

export interface MarkAllMaritimeNotificationsReadResult {
  port: number
  shipment: number
  total: number
}

export async function markAllMaritimeNotificationsRead(): Promise<MarkAllMaritimeNotificationsReadResult> {
  return api<MarkAllMaritimeNotificationsReadResult>(
    '/api/v1/maritime-alerts/notifications/read-all',
    { method: 'POST' },
  )
}
